src/subscript-loc.c

```diff
@@ -385,31 +436,52 @@ r_obj* chr_as_location(r_obj* subscript,
     r_abort("`names` must be a character vector.");
   }
 
+  bool remove_missing = false;
+
   r_obj* matched = KEEP(Rf_match(names, subscript, r_globals.na_int));
+  r_attrib_poke_names(matched, r_names(subscript));
 
   r_ssize n = r_length(matched);
-  const int* p = r_int_cbegin(matched);
+  int* p = r_int_begin(matched);
   r_obj* const * ip = r_chr_cbegin(subscript);
 
   for (r_ssize k = 0; k < n; ++k) {
-    if (p[k] != r_globals.na_int) {
-      continue;
+    const r_obj* elt = ip[k];
+
+    if (elt == r_strs.empty) {
+      // `""` never matches, even if `names` contains a `""` name
+      stop_subscript_empty(subscript, opts);
+    }
```

Comment:
i.e. if the user supplies `""` as an index in `vec_as_location()`, that is unconditionally an error because it can't match anything (#1489)